src/udev/udev-node.c

```diff
@@ -310,25 +240,28 @@ static int stack_directory_update(sd_device *dev, int fd, bool add) {
                 }
         }
 
-        r = update_timestamp(dev, fd, &st);
-        if (r < 0)
-                return r;
-
-        return 0;
+        return 1; /* Updated. */
 }
 
-static int stack_directory_open(const char *dirname) {
+static int stack_directory_lock(const char *dirname) {
         _cleanup_close_ int fd = -1;
         int r;
 
         assert(dirname);
 
-        r = mkdir_p(dirname, 0755);
+        r = mkdir_parents(dirname, 0755);
         if (r < 0)
                 return r;
 
-        fd = open(dirname, O_CLOEXEC | O_DIRECTORY | O_NOFOLLOW | O_RDONLY);
+        fd = open_mkdir_at(AT_FDCWD, dirname, O_CLOEXEC | O_DIRECTORY | O_NOFOLLOW | O_RDONLY, 0700);
         if (fd < 0)
+                return fd;
+
+        /* To prevent an arbitrary processe locks the directory. */
+        if (fchmod(fd, 0700) < 0)
```

Comment:
given you already specified 0700 during mkdir creation above this line is actually unnecessary...

Reply:
(but the comment is good, and should stay)

Reply:
The directory may be created by an old udevd. In that case, we need to explicitly update the access mode, as open_mkdir_at() does not change the access mode for the existing directory.

Reply:
true, but in that case you are actually vulnerable, because bad code might have locked the dir beforehand...

not sure it's worth really thinking about that too much, but to me it sounds pointless trying to address a case that is flawed anyway...

Reply:
Hmm, indeed... But, if the directory was not locked, then we should avoid any further lock by unprivileged processes, right?

Reply:
well, i mean, it doesn't properly fix the problem, but it certainly makes it a bit better....

maybe add a comment `/* Try to lock down the access mode to cover for upgrades, even though this won't kick out clients who already locked the dir in case it was world-readable as it was historically */`

Reply:
> Hmm, shouldn't we just use a stack directory with a different name to avoid the locking issues?

Now, ".lock" file is created under the directory. Then, I think it is not necessary to rename the directory.

Reply:
ok, but then also remove the this fchmod line, as we won't need it anymore